### PR TITLE
James 2682 Prevent infinite loop with Sieve error notifications

### PR DESCRIPTION
**Pull request by pleira**

### CODE REVIEW

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java [hunk 1]
```diff
@@ -114,7 +119,14 @@ private SieveFactory createFactory(Log log) throws MessagingException {
     public boolean execute(MailAddress recipient, Mail mail) throws MessagingException {
         Preconditions.checkNotNull(recipient, "Recipient for mail to be spooled cannot be null.");
         Preconditions.checkNotNull(mail.getMessage(), "Mail message to be spooled cannot be null.");
-
+        Optional<Attribute> oSieveNotification = mail.getAttribute(SIEVE_NOTIFICATION);
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java]: Maybe 

```
boolean isSieveNotification  = AttributeUtils.getValueAndCast(mail, SIEVE_NOTIFICATION, Boolean.class).orElse(false);
if (isSieveNotification) {
     // logic
}
```

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java [hunk 2]
```diff
@@ -114,7 +119,14 @@ private SieveFactory createFactory(Log log) throws MessagingException {
...
+        if (oSieveNotification.isPresent()) {
+            AttributeValue<Boolean> attrValue = (AttributeValue<Boolean>) oSieveNotification.get().getValue();
+            if (attrValue.value()) {
+                throw new MessagingException("Do not process Sieve error notification emails");
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java]: I would rather `return false;`

After all, the sieve notifications are valid messages we should handle.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/SieveIntegrationTest.java [hunk 3]
```diff
@@ -937,6 +938,15 @@ public void vacationDaysAddressesShouldWork() throws Exception {
             .build();
         assertThat(fakeMailContext.getSentMails()).containsExactly(expectedSentMail);
     }
+    
+    @Test(expected = MessagingException.class)
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/delivery/SieveIntegrationTest.java]: No we don't expect a MessagingException.

We want the Sieve mailet to perform normally IE do nothing, upon sieve notification, isn't it? 

```
assertThat(mail.getAttribute(ATTRIBUTE_NAME)).isEmpty()
```

Should be enough to ensure Sieve was not executed...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java [hunk 4]
```diff
@@ -121,10 +120,7 @@ public boolean execute(MailAddress recipient, Mail mail) throws MessagingExcepti
         Preconditions.checkNotNull(recipient, "Recipient for mail to be spooled cannot be null.");
         Preconditions.checkNotNull(mail.getMessage(), "Mail message to be spooled cannot be null.");
         boolean isSieveNotification = AttributeUtils.getValueAndCastFromMail(mail, SIEVE_NOTIFICATION, Boolean.class).orElse(false);
-        if (isSieveNotification) {
-           	return false;
-        }
-        return sieveMessage(recipient, mail);
+        return !isSieveNotification ? sieveMessage(recipient, mail) : false;
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java]: The previous version was more explicit.

`4 spaces` indent is all of what is needed to fix the checkstyle, no?

**Comment by chibenwa** [comment]
> don't you think an integration test would be interesting?

That would involve on a MIME4J parsing issue that might get eventually solved? What would be the value of such a test?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java (hunk 4 above)
Comment (rouazana) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/delivery/SieveExecutor.java]: I agree, we tend to not use the ternary operation in the code base.

**Comment by chibenwa** [comment]
:green_apple:  before CI restart

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Hi,

This pull request has just been merged.

Thank you again for your dedication toward the James project. 

Cheers,

Benoit

**Pull request closed without merging**